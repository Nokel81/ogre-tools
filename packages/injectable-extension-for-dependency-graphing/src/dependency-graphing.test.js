import {
  dependencyGraphCustomizerToken,
  plantUmlDependencyGraphInjectable,
  registerDependencyGraphing,
} from './dependency-graphing';

import isEqual from 'lodash/fp/isEqual';
import {
  createContainer,
  getInjectable,
  getInjectionToken,
  lifecycleEnum,
} from '@ogre-tools/injectable';

const getDi = (...injectables) => {
  const di = createContainer();

  injectables.forEach(di.register);

  return di;
};

describe('createContainer.dependency-graph', () => {
  it('given dependency graphing, dependencies and injected, creates Plant-UML graph', async () => {
    const parentInjectable = getInjectable({
      id: 'some-parent-injectable',

      instantiate: async di => {
        // Inject same injectable twice for coverage
        di.inject(syncChildInjectable);
        di.inject(syncChildInjectable);
        await di.inject(asyncChildInjectable);
        di.inject(keyedInjectable);
      },
    });

    const syncChildInjectable = getInjectable({
      id: 'some-sync-child-injectable',
      instantiate: di => di.injectMany(injectionToken),
      lifecycle: lifecycleEnum.transient,
    });

    const asyncChildInjectable = getInjectable({
      id: 'some-async-child-injectable',
      instantiate: async di => {
        di.inject(customizableAsyncInjectable);

        return 'irrelevant';
      },
    });

    const keyedInjectable = getInjectable({
      id: 'some-keyed-injectable',
      instantiate: () => 'irrelevant',
      lifecycle: lifecycleEnum.keyedSingleton({
        getInstanceKey: () => 'irrelevant',
      }),
    });

    const injectionToken = getInjectionToken({ id: 'some-injection-token' });

    const tokenInjectable = getInjectable({
      id: 'some-token-injectable',
      instantiate: () => 'irrelevant',
      injectionToken,
    });

    const customizableSyncInjectable = getInjectable({
      id: 'some-customizable-sync-injectable',
      instantiate: () => 'some-customizable-instance',
      injectionToken,
    });

    const customizableAsyncInjectable = getInjectable({
      id: 'some-customizable-async-injectable',
      instantiate: async () => 'some-customizable-instance',
    });

    const graphCustomizer = getInjectable({
      id: 'some-dependency-graph-customizer',

      instantiate: () => ({
        shouldCustomize: isEqual('some-customizable-instance'),

        customizeLink: link => {
          link.infos.add('some-custom-link-info');
          link.lineColor = 'orange';
          link.textColor = 'green';
        },

        customizeNode: node => {
          node.infos.add('some-custom-node-info');
        },
      }),

      decorable: false,

      injectionToken: dependencyGraphCustomizerToken,
    });

    const setuppable = getInjectable({
      id: 'some-setuppable',
      instantiate: () => 'irrelevant',

      setup: async di => {
        await di.inject(syncChildInjectable);
        await di.inject(setuppable);
      },
    });

    const di = getDi(
      parentInjectable,
      syncChildInjectable,
      asyncChildInjectable,
      tokenInjectable,
      setuppable,
      keyedInjectable,
      customizableSyncInjectable,
      customizableAsyncInjectable,
      graphCustomizer,
    );

    registerDependencyGraphing(di);

    await di.runSetups();

    await di.inject(parentInjectable);

    const graph = di.inject(plantUmlDependencyGraphInjectable);

    expect(graph).toMatchSnapshot();
  });
});
