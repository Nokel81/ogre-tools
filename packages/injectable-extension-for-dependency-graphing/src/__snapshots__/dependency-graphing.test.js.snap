// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createContainer.dependency-graph given dependency graphing, dependencies and injected, creates Plant-UML graph 1`] = `
"@startuml
hide members
class \\"some-token-injectable\\" as someTokenInjectable<Singleton> << (S,lightGreen) >> $some-token-injectable $some-parent-injectable $some-sync-child-injectable $some-injection-token #line:darkRed
class \\"some-customizable-sync-injectable\\" as someCustomizableSyncInjectable<Singleton\\\\nsome-custom-node-info> << (S,lightGreen) >> $some-customizable-sync-injectable $some-parent-injectable $some-sync-child-injectable $some-injection-token #line:darkRed
class \\"some-injection-token\\" as someInjectionToken<Transient\\\\nToken> << (X,orange) >> $some-injection-token $some-parent-injectable $some-sync-child-injectable #line:green
class \\"some-sync-child-injectable\\" as someSyncChildInjectable<Transient> << (T,orchid) >> $some-sync-child-injectable $some-parent-injectable #line:darkRed
class \\"some-customizable-async-injectable\\" as someCustomizableAsyncInjectable<Singleton\\\\nAsync\\\\nsome-custom-node-info> << (S,lightGreen) >> $some-customizable-async-injectable $some-parent-injectable $some-async-child-injectable #line:darkRed
class \\"some-async-child-injectable\\" as someAsyncChildInjectable<Singleton\\\\nAsync> << (S,lightGreen) >> $some-async-child-injectable $some-parent-injectable #line:darkRed
class \\"some-parent-injectable\\" as someParentInjectable<Singleton\\\\nAsync> << (S,lightGreen) >> $some-parent-injectable #line:darkRed
class \\"some-keyed-injectable\\" as someKeyedInjectable<Keyed> << (K,pink) >> $some-keyed-injectable $some-parent-injectable #line:darkRed
someInjectionToken --[#black,plain,thickness=1]up* someTokenInjectable #text:black  
someInjectionToken --[#orange,plain,thickness=1]up* someCustomizableSyncInjectable #text:green  : some-custom-link-info 
someSyncChildInjectable --[#black,plain,thickness=1]up* someInjectionToken #text:black  
someParentInjectable --[#black,plain,thickness=1]up* someSyncChildInjectable #text:black  
someAsyncChildInjectable --[#orange,plain,thickness=4]up* someCustomizableAsyncInjectable #text:green  : Async\\\\nsome-custom-link-info 
someParentInjectable --[#black,plain,thickness=4]up* someAsyncChildInjectable #text:black  : Async 
someParentInjectable --[#black,plain,thickness=1]up* someKeyedInjectable #text:black  
@enduml"
`;
